### PR TITLE
[Feature] Custom list prefixes for AttributedStringVisitor

### PR DESCRIPTION
**Pull request by dloic**
After investigation, I've tried to find the least intrusive way to add the feature of custom list prefixes for AttributedStringVisitor. I've decided to go for ListItemPrefixgenerator changes, as fitting the feature in the styler was not possible.

The PR is not changing the current behavior, but opening a bit the API so one can provide another implementation. See below for an example:

**Definition of a custom ListPrefixGenerator:**

```swift
public class CustomListItemPrefixGeneratorBuilder: ListItemPrefixGeneratorBuilder {
    public func build(listType: List.ListType, numberOfItems: Int, nestDepth: Int) -> ListItemPrefixGenerator {
        return CustomListItemPrefixGenerator(listType: listType, numberOfItems: numberOfItems, nestDepth: nestDepth)
    }
    
}

public class CustomListItemPrefixGenerator: ListItemPrefixGenerator {

    private var prefixes: IndexingIterator<[String]>

    required public init(listType: List.ListType, numberOfItems: Int, nestDepth: Int) {

        switch listType {
        case .bullet:
            let prefix = CustomListItemPrefixGenerator.unorderedPrefix(nestDepth: nestDepth)
            prefixes = [String](repeating: prefix, count: numberOfItems)
                .makeIterator()

        case .ordered(let start):
            prefixes = (start..<(start + numberOfItems))
                .map { CustomListItemPrefixGenerator.orderedPrefix(value: $0, nestDepth: nestDepth) }
                .makeIterator()
        }
    }

    public func next() -> String? {
        prefixes.next()
    }

    private static func unorderedPrefix(nestDepth: Int) -> String {
        switch nestDepth {
        case 0:
            return "●"
        case 1:
            return "○"
        case _ where nestDepth >= 2:
            return "▪︎"
        default:
            return "●"
        }
    }

    private static func orderedPrefix(value: Int, nestDepth: Int) -> String {
        switch value {
        case 1:
            return "1️⃣"
        case 2:
            return "2️⃣"
        case 3:
            return "3️⃣"
        default:
            return "🔢"
        }
    }
}
```

**Usage:**
```swift
let document = try! markdown.toDocument(.default)
let visitor = AttributedStringVisitor(
    styler: DownStyler(),
    options: .default,
    listPrefixGeneratorBuilder: CustomListItemPrefixGeneratorBuilder()
)
subview.attributedText = document.accept(visitor)
```

**Render:**
Capture d’écran, le 2021-04-19 à 15 47 27

### CODE REVIEW

**Comment by dloic** [comment]
Hi @johnxnguyen! Not sure who should be reviewing this one :)

I've checked the Travis CI build error, seems related to log file length, I cannot see the error. Do you confirm?

**Comment by johnxnguyen** [comment]
Hey @dloic , thanks for the PR! I'll try to take a look at this tonight and get back to you soon.

**Comment by dloic** [comment]
@johnxnguyen thanks a lot!

File: Sources/Down/AST/Visitors/AttributedStringVisitor.swift [hunk 1]
```diff
@@ -20,17 +20,23 @@ public class AttributedStringVisitor {
 
     private let styler: Styler
     private let options: DownOptions
+    private let listPrefixGeneratorBuilder: ListItemPrefixGeneratorBuilder
```
Comment (johnxnguyen) [inline Sources/Down/AST/Visitors/AttributedStringVisitor.swift:23]: Perhaps `ListItemPrefixGeneratorBuilder` is not necessary if instead we use a closure:

```swift
private let listPrefixGeneratorBuilder: (List) -> ListItemPrefixGenerator
```

What do you think?

File: Sources/Down/AST/Visitors/AttributedStringVisitor.swift [hunk 2]
```diff
@@ -20,17 +20,23 @@ public class AttributedStringVisitor {
...
     /// - parameters:
     ///     - styler: used to style the markdown elements.
     ///     - options: may be used to modify rendering.
+    ///     - listPrefixGeneratorBuilder: may be used to modify list prefixes.
 
-    public init(styler: Styler, options: DownOptions = .default) {
+    public init(
+        styler: Styler,
+        options: DownOptions = .default,
+        listPrefixGeneratorBuilder: ListItemPrefixGeneratorBuilder = StaticListItemPrefixGeneratorBuilder()) {
```
Comment (johnxnguyen) [inline Sources/Down/AST/Visitors/AttributedStringVisitor.swift:36]: For formatting consistency:

```suggestion
    public init(
        styler: Styler,
        options: DownOptions = .default,
        listPrefixGeneratorBuilder: ListItemPrefixGeneratorBuilder = StaticListItemPrefixGeneratorBuilder()
    ) {
```

File: Sources/Down/AST/Visitors/AttributedStringVisitor.swift [hunk 3]
```diff
@@ -53,7 +59,12 @@ extension AttributedStringVisitor: Visitor {
     }
 
     public func visit(list node: List) -> NSMutableAttributedString {
-        listPrefixGenerators.append(ListItemPrefixGenerator(list: node))
+
+        listPrefixGenerators.append(
+            listPrefixGeneratorBuilder.build(
+            listType: node.listType,
+            numberOfItems: node.numberOfItems,
+            nestDepth: node.nestDepth))
```
Comment (johnxnguyen) [inline Sources/Down/AST/Visitors/AttributedStringVisitor.swift:67]: If you create a convenience init that accepts the list node (and possibly use a closure as mentioned above), you could simplify this to:

```swift
listPrefixGenerators.append(listPrefixGeneratorBuilder(node))
```

File: Sources/Down/AST/Visitors/ListItemPrefixGenerator.swift [hunk 4]
```diff
@@ -8,34 +8,43 @@
...
 
-class ListItemPrefixGenerator {
+/// A ListItemPrefixGenerator is an object used to generate list item prefix.
+///
+/// ListItemPrefixGenerator are created by `ListItemPrefixGeneratorBuilder`, the later used in
+/// conjunction with an instance of `AttributedStringVisitor`.
+public protocol ListItemPrefixGenerator {
+    init(listType: List.ListType, numberOfItems: Int, nestDepth: Int)
+    func next() -> String?
+}
```
Comment (johnxnguyen) [inline Sources/Down/AST/Visitors/ListItemPrefixGenerator.swift:18]: You could create a convenience initializer as a protocol extension that accepts the `List` node as a single argument, which would clean up the call site a fair bit.

**Review by johnxnguyen — COMMENTED** [review]
Hey 👋🏻  Sorry to keep you waiting, I had quite a busy week.

Generally it looks good to me! I just have a few suggestions to tidy it up a little. Let me know what you think.

**Comment by johnxnguyen** [comment]
> I've checked the Travis CI build error, seems related to log file length, I cannot see the error. Do you confirm?

@dloic hmm apparently Travis enforces a 4MB log length limit, it's the first time I've seen this issue though. Try to add the `- quiet` option at the end of the three `xcodebuild` commands in the travis config file, that should shrink the log output sufficiently.

File: Sources/Down/AST/Visitors/AttributedStringVisitor.swift (hunk 1 above)
Comment (dloic) [reply Sources/Down/AST/Visitors/AttributedStringVisitor.swift:23]: I was not a fan of the builder neither. Will refactor based on your suggestion.

File: Sources/Down/AST/Visitors/AttributedStringVisitor.swift (hunk 2 above)
Comment (dloic) [reply Sources/Down/AST/Visitors/AttributedStringVisitor.swift:36]: Will do!

**Comment by dloic** [comment]
@johnxnguyen Thanks for the review. Branch updated taking into account all your feedback.
I did also tweak the travis script as suggested. Let see the result!

Updated usage of the new feature with _builder_ as closure:

```swift
let visitor = AttributedStringVisitor(
    styler: DownStyler(),
    options: .default,
    listPrefixGeneratorBuilder: { CustomListItemPrefixGenerator(list: $0) }
)
```

**Review by johnxnguyen — APPROVED** [review]
Looks good to me 👍🏻 Thanks @dloic for your contribution!

**Comment by dloic** [comment]
Thanks to you @johnxnguyen for the review and approval! There is no rush, but I was wondering if you plan to do a new release for the new changes.

**Comment by johnxnguyen** [comment]
I will draft a new release likely by the end of this week.

**Comment by dloic** [comment]
Perfect, thanks a lot!

**Comment by johnxnguyen** [comment]
FYI @dloic this has been release. Thanks again for the contribution!

**Pull request merged**